corehq/motech/repeaters/models.py

```diff
@@ -1120,7 +1131,31 @@ def attempts(self):
 
     @property
     def num_attempts(self):
-        return self.sqlrepeatrecordattempt_set.count()
+        # Uses `len(queryset)` instead of `queryset.count()` to use
+        # prefetched attempts, if available.
+        return len(self.attempts)
+
+    def get_numbered_attempts(self):
+        for i, attempt in enumerate(self.attempts):
+            yield i + 1, attempt
+
+    @property
+    def record_id(self):
+        # Used by Repeater.get_url() ... by SQLRepeatRecordReport._make_row()
+        return self.pk
+
+    @property
+    def failure_reason(self):
+        if has_failed(self):
+            return self.last_message
+        else:
+            return ''
+
+    @property
+    def last_message(self):
+        # Uses `list(queryset)[-1]` instead of `queryset.last()` to use
+        # prefetched attempts, if available.
+        return list(self.attempts)[-1].message
```

Comment:
What do you think of combining these two methods since `failure_reason` is the only thing that calls `last_message`, and `last_message` would raise `IndexError` if it were called when there were no attempts? I think `IndexError` is impossible here, but exposing it as a public method tempts other future/unsuspecting code to call it in less certain circumstances.
```suggestion
        # Uses `list(queryset)[-1]` instead of `queryset.last()` to use
        # prefetched attempts, if available.
        attempts = list(self.attempts)
        return attempts[-1].message if has_failed(self) else ''
```

Reply:
In two PR's time it will also get called [over here](https://github.com/dimagi/commcare-hq/blob/212756c8d183221e17b304e11cf2bbc4c8c8b9f9/corehq/motech/repeaters/management/commands/update_cancelled_records.py#L133-L134). But I agree that it should never raise an IndexError.

Reply:
94a1b0ae5b6